src/native/watchdog/watchdog.cpp

```diff
@@ -0,0 +1,128 @@
+#include <cstdio>
+#include <cstdlib>
+#include <cstdarg>
+#include <errno.h>
+#include <signal.h>
+
+#ifdef _WIN32
+#include <windows.h>
+#include <string>
+#else
+#include <unistd.h>
+#include <sys/wait.h>
+#endif
+
+int run_timed_process(const long, const int, const char *[]);
+
+int main(const int argc, const char *argv[])
+{
+    if (argc < 3)
+    {
+        printf("There are missing arguments. Got %d instead of 3+ :(\n", argc);
+        return EXIT_FAILURE;
+    }
+
+    const long timeout_ms = strtol(argv[1], nullptr, 10);
+    int exit_code = run_timed_process(timeout_ms, argc-2, &argv[2]);
+
+    printf("App Exit Code: %d\n", exit_code);
+    return EXIT_SUCCESS;
+}
+
+int run_timed_process(const long timeout, const int exe_argc, const char *exe_path_and_argv[])
+{
+#ifdef _WIN32
+    std::string cmdline(exe_path_and_argv[0]);
+
+    for (int i = 1; i < exe_argc; i++)
+    {
+        cmdline.append(" ");
+        cmdline.append(exe_path_and_argv[i]);
+    }
+
+    STARTUPINFO startup_info;
+    PROCESS_INFORMATION proc_info;
+    int exit_code;
+
+    ZeroMemory(&startup_info, sizeof(startup_info));
+    startup_info.cb = sizeof(startup_info);
+    ZeroMemory(&proc_info, sizeof(proc_info));
+
+    if (!CreateProcess(NULL, &cmdline[0], NULL, NULL, FALSE, 0, NULL, NULL,
+                       &startup_info, &proc_info))
+    {
+        int error_code = GetLastError();
+        printf("Process creation failed... Code %d.\n", error_code);
+        return error_code;
+    }
+
+    WaitForSingleObject(proc_info.hProcess, timeout);
+    GetExitCodeProcess(proc_info.hProcess, &exit_code);
+
+    CloseHandle(proc_info.hProcess);
+    CloseHandle(proc_info.hThread);
+    return exit_code;
+
+#else
+    const int check_interval = 1000;
+    int check_count = 0;
+    char **args = new char *[exe_argc];
```

Comment:
Consider a vector<string> here. You forgot to delete

Reply:
I forgot the deletion, thanks for pointing it out Andy. The reason we're using a `char*[]` is because that's the type that `execv()` requires.

Reply:
`unique_ptr<char*[]>` then?

Reply:
That might work as well. I'll try it.